provider/ec2/instancetypes.go

```diff
@@ -0,0 +1,328 @@
+	"arm":     arch.ARM,
+	"aarch64": arch.ARM64,
+}
+
+func archName(in string) string {
+	if archName, ok := archNames[in]; ok {
+		return archName
+	}
+	return in
+}
+
+func (e *environ) supportedInstanceTypes(ec2Session ec2iface.EC2API, ctx context.ProviderCallContext) (result []instances.InstanceType, err error) {
+	e.instTypesMutex.Lock()
+	defer e.instTypesMutex.Unlock()
+
+	defer func() {
+		if err == nil {
+			if len(e.instTypes) == 0 {
+				logger.Tracef("got instance types:\n%#v", result)
+			}
+			e.instTypes = result
+		}
+	}()
+
+	// Use a cached copy if populated as it's mildly
+	// expensive to fetch each time.
+	if len(e.instTypes) > 0 {
+		return e.instTypes, nil
+	}
+
+	// First get all the zone names for the current region.
+	zoneResults, err := ec2Session.DescribeAvailabilityZones(&ec2.DescribeAvailabilityZonesInput{
+		Filters: []*ec2.Filter{{
+			Name:   aws.String("region-name"),
+			Values: []*string{aws.String(e.cloud.Region)},
+		}},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var zoneNames []string
+	zoneFilter := &ec2.Filter{Name: aws.String("location")}
+	for _, z := range zoneResults.AvailabilityZones {
+		zoneNames = append(zoneNames, *z.ZoneName)
+		zoneFilter.Values = append(zoneFilter.Values, z.ZoneName)
+	}
+
+	// Query the instance type names for the region and credential in use.
+	var instTypeNames []*string
+	instanceTypeRegions := make(map[string]set.Strings)
+	var token *string
+	for {
+		offeringResults, err := ec2Session.DescribeInstanceTypeOfferings(&ec2.DescribeInstanceTypeOfferingsInput{
+			LocationType: aws.String("availability-zone"),
+			MaxResults:   aws.Int64(100),
+			NextToken:    token,
+			Filters:      []*ec2.Filter{zoneFilter},
+		})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, offering := range offeringResults.InstanceTypeOfferings {
+			if _, ok := instanceTypeRegions[*offering.InstanceType]; !ok {
+				instanceTypeRegions[*offering.InstanceType] = set.NewStrings()
+			}
+			instanceTypeRegions[*offering.InstanceType].Add(*offering.Location)
+			instTypeNames = append(instTypeNames, offering.InstanceType)
+		}
+		token = offeringResults.NextToken
+		if token == nil {
+			break
+		}
+	}
+
+	// Populate the costs fo the instance types in use.
+	costs := make(map[string]uint64)
+	if err := instanceTypeCosts(ec2Session, instTypeNames, zoneNames, costs); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Compose the results.
+	var allInstanceTypes []instances.InstanceType
+	for {
+		querySize := len(instTypeNames)
+		if querySize > 100 {
+			querySize = 100
+		}
+		instTypeParams := &ec2.DescribeInstanceTypesInput{
+			InstanceTypes: instTypeNames[0:querySize],
+		}
+		instTypeResults, err := ec2Session.DescribeInstanceTypes(instTypeParams)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, info := range instTypeResults.InstanceTypes {
+			instType := instances.InstanceType{
+				Name:       *info.InstanceType,
```

Comment:
nil check needed

Reply:
InstanceType is never nil

Reply:
Defensive programming would ignore the assumption and just continue the loop if the value is nil.

Reply:
added check here and in other places eg zone names